### ISSUE TITLE
MAE (and  Logcosh / Adaptive Huber) as possible loss functions to include

### ISSUE THREAD

**Issue by BradKML** [perpetual-ml/perpetual#46]
Sometimes there are data that requires different loss functions. e.g. MSE are more sensitive to extreme values, while MAE are more robust to those same values. LogCosh (low compute) and Adaptive Huber (high compute) makes compromises between the two sides.

**Comment by deadsoul44** [comment]
We're planning to expand the range of loss functions our algorithm supports. To help us prioritize, please let us know which functions are most important to you and why.

**Comment by deadsoul44** [comment]
What do you mean by Adaptive Huber? I couldn't find much resources about it.

**Comment by BradKML** [comment]
I was given two examples for this. There are enough references to "Adaptive Huber Regression" but I think the examples below are a good starting point.

```
import numpy as np

def adaptive_huber_loss(y_true, y_pred, k=1.345):
    """
    Adaptive Huber Loss function.

    Args:
        y_true (np.ndarray): Array of true values.
        y_pred (np.ndarray): Array of predicted values.
        k (float): Constant factor for MAD, e.g., 1.345 for consistency with std dev
                  for Gaussian noise.  Higher values make it more robust.

    Returns:
        np.ndarray: Array of loss values.
    """
    residuals = np.abs(y_true - y_pred)
    delta = k * np.median(residuals)

    loss = np.where(
        residuals <= delta,
        0.5 * residuals**2,
        delta * residuals - 0.5 * delta**2
    )

    return loss

# Example Usage:
if __name__ == '__main__':
    y_true = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 100.0])  # Outlier included
    y_pred = np.array([1.2, 1.8, 2.8, 4.2, 4.8, 5.0])

    loss_values = adaptive_huber_loss(y_true, y_pred)
    print(f"Adaptive Huber Loss values: {loss_values}")
    print(f"Mean Adaptive Huber Loss: {np.mean(loss_values)}")

```

```
import torch

def adaptive_huber_loss(y_true, y_pred, alpha=0.9):
    """
    Adaptive Huber loss with automatically determined delta.
    
    Args:
        y_true: Ground truth values
        y_pred: Predicted values
        alpha: Quantile for delta calculation (default: 0.9)
    
    Returns:
        Mean loss value
    """
    residual = y_true - y_pred
    abs_residual = torch.abs(residual)
    
    # Calculate delta as the quantile of absolute residuals
    delta = torch.quantile(abs_residual.detach(), alpha)
    
    # Apply Huber loss with the adaptive delta
    quadratic_mask = abs_residual <= delta
    linear_mask = ~quadratic_mask
    
    loss = torch.zeros_like(residual)
    loss[quadratic_mask] = 0.5 * residual[quadratic_mask]**2
    loss[linear_mask] = delta * (abs_residual[linear_mask] - 0.5 * delta)
    
    return torch.mean(loss)
```
